<h1 style="color: #9f43c3ff; border-bottom: 3px solid #9f43c3ff; padding-bottom: 8px;">
AI & ML Course with BinX Week 3 Day 4 : Trees, Forests, SVMs & k-NN
</h1>

<blockquote style="border-left: 3px solid #2e1a9aff; padding-left: 12px; margin-left: 0;">

<b>Day 4 Learning Objectives:</b>
- 📂 Load Telco Customer Churn dataset (`Customer-Churn.csv`) and handle missing values & categorical features.
- 🎯 Define Feature Matrix ($X$) and Target Vector ($y$) using one-hot encoding (`pd.get_dummies`).
- 🤖 Train four distinct classifiers: Decision Tree, Random Forest, Support Vector Machine (SVM), and k-Nearest Neighbors (k-NN) on an 80/20 train-test split.
- 📊 Compare model performance across all four classifiers using F1-score on the exact same dataset split.
- 🌲 Report and interpret the top feature importances from the Random Forest model.
- 🏆 Identify the best-performing model for this dataset and justify why it achieved superior performance.

</blockquote>


## <span style="color: #309c42ff">4.1 Environment Setup & Library Imports</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We begin by importing core numerical, data manipulation, visualization, and Scikit-Learn tree, ensemble, vector machine, and neighbor classifiers alongside evaluation metrics.

</blockquote>


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import f1_score


print("All required libraries successfully imported.")

All required libraries successfully imported.


## <span style="color: #309c42ff">4.2 Dataset Loading & Exploration</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We load the Telco Customer Churn dataset from <code>../../Data/Customer-Churn.csv</code> and inspect dataset dimensions, data types, and initial target class distribution.

</blockquote>


In [2]:
df = pd.read_csv("../../Data/Customer-Churn.csv")

print(f"Dataset Shape: {df.shape[0]} rows, {df.shape[1]} columns\n")
print("Target Class Distribution (Churn):")
print(df['Churn'].value_counts(normalize=True).map("{:.2%}".format))

df.info()

Dataset Shape: 7043 rows, 21 columns

Target Class Distribution (Churn):
Churn
No     73.46%
Yes    26.54%
Name: proportion, dtype: str
<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMo

## <span style="color: #309c42ff">4.3 Data Preprocessing & Feature Matrix Construction</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We convert <code>TotalCharges</code> to numerical values, impute missing values using the median, convert the target variable <code>Churn</code> into a binary format, and perform one-hot encoding (`pd.get_dummies`) to build Feature Matrix ($X$) and Target Vector ($y$).

</blockquote>


In [3]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df['TotalCharges'] = df['TotalCharges'].fillna(df['TotalCharges'].median())

df['Churn_Binary'] = (df['Churn'] == 'Yes').astype(int)

X = pd.get_dummies(df.drop(columns=['customerID', 'Churn', 'Churn_Binary']), drop_first=True)
y = df['Churn_Binary']

print(f"Feature Matrix X Shape: {X.shape}")
print(f"Target Vector y Shape: {y.shape}")
print(f"Total encoded features: {X.shape[1]}")

Feature Matrix X Shape: (7043, 30)
Target Vector y Shape: (7043,)
Total encoded features: 30


## <span style="color: #309c42ff">4.4 Model Training (Decision Tree, Random Forest, SVM & k-NN)</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We partition the dataset into an 80% training set and a 20% testing set (`random_state=42`, `stratify=y`). We then initialize and fit four distinct classifiers (`DecisionTreeClassifier`, `RandomForestClassifier`, `SVC`, and `KNeighborsClassifier`) on the exact same training split.

</blockquote>


In [4]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

dt = DecisionTreeClassifier(max_depth=5, random_state=42)
dt.fit(X_train, y_train)

rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

svm = SVC(kernel="rbf", probability=True, random_state=42)
svm.fit(X_train, y_train)

knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

print("✅ Models successfully trained.")
print(f"Training samples: {len(X_train)} | Test samples: {len(X_test)}")

c:\Users\Amr-Sheqwara\Desktop\AI_ML_Internship\.venv\Lib\site-packages\sklearn\svm\_base.py:239: FutureWarning: The `probability` parameter was deprecated in 1.9 and will be removed in version 1.11. Use `CalibratedClassifierCV(SVC(), ensemble=False)` instead of `SVC(probability=True)`
  warnings.warn(


✅ Models successfully trained.
Training samples: 5634 | Test samples: 1409


## <span style="color: #309c42ff">4.5 Model Evaluation & Comparison</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

Here we evaluate all four models on the same test set using the F1-score metric and assemble the performance results into a single unified comparison table.

</blockquote>


In [5]:
y_pred_dt = dt.predict(X_test)
y_pred_rf = rf.predict(X_test)
y_pred_svm = svm.predict(X_test)
y_pred_knn = knn.predict(X_test)

f1_dt = f1_score(y_test, y_pred_dt)
f1_rf = f1_score(y_test, y_pred_rf)
f1_svm = f1_score(y_test, y_pred_svm)
f1_knn = f1_score(y_test, y_pred_knn)

results = pd.DataFrame({
    'Model': ['Decision Tree', 'Random Forest', 'k-NN', 'SVM'],
    'F1-Score': [f1_dt, f1_rf, f1_knn, f1_svm]
})

print(results)


           Model  F1-Score
0  Decision Tree  0.584527
1  Random Forest  0.550075
2           k-NN  0.493865
3            SVM  0.000000


## <span style="color: #309c42ff">4.6 Random Forest Feature Importances</span>

<blockquote style="border-left: 3px solid #FD1D1D; padding-left: 12px; margin-left: 0;">

We extract and rank the feature importances (`model.feature_importances_`) from the Random Forest classifier to identify which customer attributes drive churn predictions most strongly.

</blockquote>


In [6]:
importances = pd.Series(rf.feature_importances_, index=X.columns).sort_values(ascending=False)

print("Top 10 Feature Importances:")
print(importances.head(10), "\n")

Top 10 Feature Importances:
TotalCharges                      0.192096
tenure                            0.174733
MonthlyCharges                    0.168413
PaymentMethod_Electronic check    0.038771
InternetService_Fiber optic       0.038641
Contract_Two year                 0.030176
gender_Male                       0.028321
OnlineSecurity_Yes                0.028191
PaperlessBilling_Yes              0.025617
Partner_Yes                       0.023326
dtype: float64 



---

## <span style="color:#F78BA0">4.7 Summary & Interpretation of Results</span>

<blockquote style="border-left: 3px solid #F78BA0; padding-left: 12px; margin-left: 0;">

📌 <b>Interpretation & Key Takeaways:</b>

- 🏆 <b>Best Performing Model:</b> <code>DecisionTreeClassifier</code> achieved the highest F1-Score (<b>0.5845</b>), outperforming Random Forest, k-NN, and SVM.
- 🌲 <b>Decision Tree vs. Random Forest:</b> The constrained single Decision Tree (`max_depth=5`) generalized better on the minority churn class than the default Random Forest ensemble (<b>0.5501</b>) on this specific unscaled dataset split.
- ⚠️ <b>Why SVM Failed (F1 = 0.0000):</b> Without feature scaling, unscaled numerical variables (`TotalCharges`, `MonthlyCharges`) completely dominated the RBF kernel distance calculations, causing the SVM to predict the majority non-churn class (`0`) for 100% of samples.
- 📉 <b>k-NN Performance (F1 = 0.4939):</b> k-NN degraded due to unscaled features and high-dimensional one-hot encoded binary columns.
- 💡 <b>Key Lesson:</b> Tree-based algorithms (Decision Trees & Random Forests) are scale-invariant and naturally handle unscaled mixed tabular features, whereas distance-based models (SVM & k-NN) fail without feature standardization.

</blockquote>

---
